### ISSUE TITLE
sql: st_distance = 0 optimiser normalisation panics

### ISSUE THREAD

**Issue by otan** [cockroachdb/cockroach#67235]
Run:

```
select ((0.0:::FLOAT8 = st_distance(st_geographyfromtext('"':::STRING::STRING)::GEOGRAPHY::GEOGRAPHY, '0105000020E610000003000000010200000006000000ACE8E56F4034564001E5B167892347C004250707FFAC60404086E88062D100402CBE02D03042624030B0A2E27001524040D4D78FD3ED1C40503EB904B04655409461920E1E6353C064F93F37E2E63C40860D80E6D9FD61C010FE5FBCDCFD4F4001020000000800000014AB992A603343C0F02E0DCABEA950C0BC0026081D554440B5FA75A07D544EC0FC44234B291A4B403ADD1E84AD9754C020A93F2B2DBC6040B8B76DC74BE530C0F22991025E3064404006137DFF38F1BFB80F136C55085B40CE6372D6537451405C0F219D02F250403070CFC13CF94640404F29A1CFD442C070D6518DF3F44540010200000002000000A87B6AC385A03BC0188C8056AF0B4F40CF7BE455759A60C0ACF151945B3B5440':::GEOGRAPHY::GEOGRAPHY, true::BOOL)::FLOAT8) AND NULL)::BOOL;
```

This panics with:

```
root@127.0.0.1:26257/defaultdb> select ((0.0:::FLOAT8 = st_distance(st_geographyfromtext('"':::STRING::STRING)::GEOGRAPHY::GEOGRAPHY, '0105000020E610000003000000010200000006000000ACE8E56F4034564001E5B167892347C004250707FFAC60404086E88062D100402CBE02D03042624030B0A2E27001524040D4D78FD3ED1C40503EB904B04655409461920E1E6353C064F93F37E2E63C40860D80E6D9FD61C010FE5FBCDCFD4F4001020000000800000014AB992A603343C0F02E0DCABEA950C0BC0026081D554440B5FA75A07D544EC0FC44234B291A4B403ADD1E84AD9754C020A93F2B2DBC6040B8B76DC74BE530C0F22991025E3064404006137DFF38F1BFB80F136C55085B40CE6372D6537451405C0F219D02F250403070CFC13CF94640404F29A1CFD442C070D6518DF3F44540010200000002000000A87B6AC385A03BC0188C8056AF0B4F40CF7BE455759A60C0ACF151945B3B5440':::GEOGRAPHY::GEOGRAPHY, true::BOOL)::FLOAT8) AND NULL)::BOOL;
ERROR: internal error: could not find overload for st_intersects
SQLSTATE: XX000
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/comp_funcs.go:124: MakeIntersectionFunction()
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:12974: ConstructEq()
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:13065: ConstructEq()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:656: constructComparison()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:286: buildScalar()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:118: buildScalar()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/project.go:176: buildProjectionList()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:960: buildSelectClause()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:879: buildSelectStmtWithoutParens()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:852: func1()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/with.go:111: processWiths()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:851: buildSelect()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:275: buildStmt()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:235: buildStmtAtRoot()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:209: Build()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:521: buildExecMemo()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:194: makeOptimizerPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:988: makeExecPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:869: dispatchToExecutionEngine()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:681: execStmtInOpenState()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:125: execStmt()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1603: func1()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1605: execCmd()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1527: run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:615: ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:645: func1()
runtime/asm_amd64.s:1371: goexit()
```

**Comment by mgartner** [comment]
We have a normalization rule in the optimizer that normalizes `st_distance(a, b) = 0` to `st_intersects(a, b)`, `FoldEqZeroSTDistance`. This normalization rule is important because it allows an inverted index to be used.

There error comes from the fact that `st_distance` accepts a third optional argument, `use_spheroid=true`, but `st_intersects` accepts only two arguments. Disabling the rule if there is three arguments is easy enough, but I think this bug reveals a more interesting issue that we need to consider.

From the [`ST_Distance` PostGIS docs](https://postgis.net/docs/ST_Distance.html):

> For geography types defaults to return the minimum geodesic distance between two geographies in meters, compute on the spheroid determined by the SRID. If use_spheroid is false, a faster spherical calculation is used.

From the [`ST_Intersects` PostGIS docs](https://postgis.net/docs/ST_Intersects.html):

> For geography, this function has a distance tolerance of about 0.00001 meters and uses the sphere rather than spheroid calculation.

**In summary, `st_distance` calculates on a spheroid by default, and on a sphere if `use_spheroid` is false. `st_intersects` calculates on the sphere always.** So it seems that this normalization rule is only valid for `GEOGRAPHY` arguments if `use_spheroid=false` (which is not the default).

Should we disable the rule for `GEOGRAPHY` arguments if `use_spheroid=false` is NOT passed? I think we have to for the sake of correctness, but I worry that very few users will be passing `use_spheroid=false`, and so few queries with filters like `st_distance(...) = 0` will be index-accelerated.

**Comment by otan** [comment]
i think i (not sure if falsely) assumed that intersection should work the same for a sphere or spheroid, but i think your answer is more technically correct and we should do what you suggest.
